### PR TITLE
[systeminfo] use timestamp for application build date from BUILDATE

### PR DESCRIPTION
**Pull request by wsnipex**
…file, if it exists

## Description

26d356dfe0dd9c5b5dcae601a0ee786a7f99aa86 introduced reading the application build date from a file. This PR expands to use this datestamp in SystemInfo screen and log.

## Motivation and Context

builds on launchpad have a long standing issue with incorrect localtime. 
This causes the SystemInfo screen to report "Compiled Nov 4 2012"
This might be wanted even where the cpp macro \_\_DATE__ reports the correct time, e.g. for reproduceable builds.

## How Has This Been Tested?

only on ubuntu

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/CompileInfo.cpp.in [hunk 1]
```diff
@@ -59,3 +59,14 @@ const char* CCompileInfo::GetCopyrightYears()
...
+
+const char* CCompileInfo::GetBuildDate()
+{
+  const std::string bdate = "@APP_BUILD_DATE@";
+  if (!bdate.empty())
+  {
+    std::string datestamp = bdate.substr(0, 4) + "-" + bdate.substr(4, 2) + "-" + bdate.substr(6, 2);
+    return datestamp.c_str();
+  }
+  return __DATE__;
```
Comment (Rechi) [inline xbmc/CompileInfo.cpp.in]: If you keep this, still no reproducible builds.

File: xbmc/utils/SystemInfo.cpp [hunk 2]
```diff
@@ -1243,7 +1243,8 @@ std::string CSysInfo::GetVersion()
 
 std::string CSysInfo::GetBuildDate()
 {
-  return StringUtils::Format("%s", __DATE__);
+  std::string bdate = CCompileInfo::GetBuildDate();
+  return StringUtils::Format("%s", bdate);
```
Comment (Rechi) [inline xbmc/utils/SystemInfo.cpp]: Why not just return the string?

File: xbmc/CompileInfo.cpp.in (hunk 1 above)
Comment (wsnipex) [reply xbmc/CompileInfo.cpp.in]: this is only used when BUILDDATE is not present

File: xbmc/utils/SystemInfo.cpp (hunk 2 above)
Comment (wsnipex) [reply xbmc/utils/SystemInfo.cpp]: dunno, the format was there before and I didn't want to mess with it

File: xbmc/CompileInfo.cpp.in (hunk 1 above)
Comment (Rechi) [reply xbmc/CompileInfo.cpp.in]: still always compiled into the binary

Comment (wsnipex) [reply xbmc/CompileInfo.cpp.in]: heh, ofc you're right. Any suggestion on how to fix this? Maybe some macro magic?

Comment (wsnipex) [reply xbmc/CompileInfo.cpp.in]: I moved it to cmake

File: xbmc/utils/SystemInfo.cpp (hunk 2 above)
Comment (wsnipex) [reply xbmc/utils/SystemInfo.cpp]: not sure what to do with the Format here, remove?

Comment (Rechi) [inline xbmc/utils/SystemInfo.cpp]: There is no need to call format, GetBuildDate is already a string.
```suggestion
  return CCompileInfo::GetBuildDate();
```

**Review by Rechi — APPROVED** [review]

**Comment by MilhouseVH** [comment]
This is how the System Info looks with this PR in LibreELEC:
s1

Note `Compiled: 1970-01-01`.

I can add [a fix](https://github.com/LibreELEC/LibreELEC.tv/commit/a3533dac82f21189baab2fcdd1d5c98ac982a216) to the LibreELEC build system, but wouldn't defaulting to the current date make sense?

**Comment by wsnipex** [comment]
works fine on ubuntu. I guess LE sets GIT_VERSION?

**Comment by MilhouseVH** [comment]
Yes, LibreELEC sets GIT_VERSION: https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/mediacenter/kodi/package.mk#L211

**Pull request merged**